File: backend/app/schemas/response.py
from typing import Literal, Union
from app.schemas.enums import AgentType
from pydantic import BaseModel, Field
from uuid import uuid4


class Message(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid4()))
    msg_type: Literal[
        "system", "agent", "user", "tool", "progress", "stream"
    ]  # system msg | agent message | user message | tool message | progress message | stream message
    content: str | None = None


class ToolMessage(Message):
    msg_type: str = "tool"
    tool_name: Literal["execute_code", "search_scholar"]
    input: dict | None = None
    output: list | None = None


class SystemMessage(Message):
    msg_type: str = "system"
    type: Literal["info", "warning", "success", "error"] = "info"


class UserMessage(Message):
    msg_type: str = "user"


class AgentMessage(Message):
    msg_type: str = "agent"
    agent_type: AgentType  # CoordinatorAgent | ModelerAgent | CoderAgent | WriterAgent


class ModelerMessage(AgentMessage):
    agent_type: AgentType = AgentType.MODELER


class CoordinatorMessage(AgentMessage):
    agent_type: AgentType = AgentType.COORDINATOR


class CodeExecution(BaseModel):
    res_type: Literal["stdout", "stderr", "result", "error"]
    msg: str | None = None


class StdOutModel(CodeExecution):
    res_type: str = "stdout"


class StdErrModel(CodeExecution):
    res_type: str = "stderr"


class ResultModel(CodeExecution):
    res_type: str = "result"
    format: Literal[
        "text",
        "html",
        "markdown",
        "png",
        "jpeg",
        "svg",
        "pdf",
        "latex",
        "json",
        "javascript",
    ]


class ErrorModel(CodeExecution):
    res_type: str = "error"
    name: str
    value: str
    traceback: str


# 代码执行结果类型
OutputItem = Union[StdOutModel, StdErrModel, ResultModel, ErrorModel]


class ProgressMessage(Message):
    """任务进度消息"""
    msg_type: str = "progress"
    type: Literal["info", "warning", "success", "error"] = "info"
    percent: float = 0.0
    phase: str = ""
    message: str = ""


class ScholarMessage(ToolMessage):
    tool_name: str = "search_scholar"
    input: dict | None = None  # query
    output: list[str] | None = None  # cites


class InterpreterMessage(ToolMessage):
    tool_name: str = "execute_code"
    input: dict | None = None  # code
    output: list[OutputItem] | None = None  # code_results


# 1. 只带 code
# 2. 只带 code result
class CoderMessage(AgentMessage):
    agent_type: AgentType = AgentType.CODER


class WriterMessage(AgentMessage):
    agent_type: AgentType = AgentType.WRITER
    sub_title: str | None = None


class StreamMessage(Message):
    """LLM 流式输出消息，前端按 message_id 拼接 delta"""
    msg_type: str = "stream"
    agent_type: str = ""       # 哪个 agent 在说话
    delta: str = ""            # 增量 token 内容
    message_id: str = ""       # 用于前端识别同一消息流
    done: bool = False         # 是否流式结束


# 所有可能的消息类型
MessageType = Union[
    SystemMessage,
    ProgressMessage,
    UserMessage,
    ModelerMessage,
    CoderMessage,
    WriterMessage,
    CoordinatorMessage,
    StreamMessage,
]

File: frontend/src/utils/response.ts
// 对应 response.py 的结构

export type SystemMessageType = 'info' | 'warning' | 'success' | 'error';

import { AgentType } from './enum';


export interface BaseMessage {
  id: string;
  msg_type: 'system' | 'agent' | 'user' | 'tool' | 'progress' | 'stream';
  content?: string | null;
}

export interface ToolMessage extends BaseMessage {
  msg_type: 'tool';
  tool_name: 'execute_code' | 'search_scholar';
  input: Record<string, unknown> | null;
  output: string[] | OutputItem[] | null;
}

export interface SystemMessage extends BaseMessage {
  msg_type: 'system';
  type: SystemMessageType;
}

export interface UserMessage extends BaseMessage {
  msg_type: 'user';
}

export interface AgentMessage extends BaseMessage {
  msg_type: 'agent';
  agent_type: AgentType;
}

export interface ModelerMessage extends AgentMessage {
  agent_type: AgentType.MODELER;
}

export interface CoordinatorMessage extends AgentMessage {
  agent_type: AgentType.COORDINATOR;
}


// 代码执行结果类型
export type ExecutionFormat =
  | 'text'
  | 'html'
  | 'markdown'
  | 'png'
  | 'jpeg'
  | 'svg'
  | 'pdf'
  | 'latex'
  | 'json'
  | 'javascript';

export interface BaseCodeExecution {
  res_type: 'stdout' | 'stderr' | 'result' | 'error';
  msg?: string;
}

export interface StdOutExecution extends BaseCodeExecution {
  res_type: 'stdout';
}

export interface StdErrExecution extends BaseCodeExecution {
  res_type: 'stderr';
}

export interface ResultExecution extends BaseCodeExecution {
  res_type: 'result';
  format: ExecutionFormat;
}

export interface ErrorExecution extends BaseCodeExecution {
  res_type: 'error';
  name: string;
  value: string;
  traceback: string;
}

export type OutputItem = StdOutExecution | StdErrExecution | ResultExecution | ErrorExecution;

export interface ScholarMessage extends ToolMessage {
  tool_name: 'search_scholar';
  input: Record<string, unknown>;
  output: string[];
}

export interface InterpreterMessage extends ToolMessage {
  tool_name: 'execute_code';
  input: {
    code: string;
  } | null;
  output: OutputItem[] | null;
}


export interface CoderMessage extends AgentMessage {
  agent_type: AgentType.CODER;
}

export interface WriterMessage extends AgentMessage {
  agent_type: AgentType.WRITER;
  sub_title?: string;
}

export interface ReviewerMessage extends AgentMessage {
  agent_type: AgentType.REVIEWER;
  review_score?: number;
  dimension_scores?: Record<string, number>;
}

export interface AnalyzerMessage extends AgentMessage {
  agent_type: AgentType.ANALYZER;
}

export interface ValidatorMessage extends AgentMessage {
  agent_type: AgentType.VALIDATOR;
}

export interface OptimizerMessage extends AgentMessage {
  agent_type: AgentType.OPTIMIZER;
}

export interface ProgressMessage extends BaseMessage {
  msg_type: 'progress';
  type: SystemMessageType;
  percent: number;
  phase: string;
  message: string;
  elapsed_time?: number;
  sub_phase?: string;
  iteration?: number;
  max_iterations?: number;
  quality_score?: number;
}

export interface TaskStatusMessage {
  id: string;
  msg_type: 'task_status';
  content?: string | null;
  status: string;
  phase?: string;
  error?: string;
}

export interface StreamMessage extends BaseMessage {
  msg_type: 'stream';
  agent_type: string;
  delta: string;
  message_id: string;
  done: boolean;
}

export type Message = SystemMessage | ProgressMessage | UserMessage | CoderMessage | WriterMessage | ModelerMessage | CoordinatorMessage | ReviewerMessage | AnalyzerMessage | ValidatorMessage | OptimizerMessage | ToolMessage | StreamMessage;